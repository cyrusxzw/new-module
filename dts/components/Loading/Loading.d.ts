import { FC } from 'react';
import type { LoadingProps } from './Loading.types';
declare const Loading: FC<LoadingProps>;
export { Loading };
