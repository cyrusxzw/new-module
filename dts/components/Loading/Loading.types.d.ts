import type { Themes } from '~/types';
declare type Sizes = 'small' | 'medium' | 'large';
declare type LoadingProps = {
  className?: string;
  isLoading: boolean;
  shouldFillSpace?: boolean;
  size?: Sizes;
  theme?: Themes;
};
export type { LoadingProps };
