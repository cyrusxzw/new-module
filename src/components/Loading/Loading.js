import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Loading.module.css';

const Loading = ({ className, isLoading, shouldFillSpace, size, theme }) => {
  if (!isLoading) return null;

  const classSet = cx(
    styles.base,
    {
      [styles.isLoading]: isLoading,
      [styles.fullSize]: shouldFillSpace,
    },
    styles[size],
    styles[theme],
    className,
  );

  return (
    <span className={classSet} data-testid="data-testid-loading">
      <span className={styles.dot} />
      <span className={styles.dot} />
      <span className={styles.dot} />
    </span>
  );
};

Loading.propTypes = {
  className: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  shouldFillSpace: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  theme: PropTypes.oneOf(['dark', 'light']),
};

Loading.defaultProps = {
  className: undefined,
  isLoading: undefined,
  shouldFillSpace: false,
  size: 'medium',
  theme: 'dark',
};

export { Loading };
