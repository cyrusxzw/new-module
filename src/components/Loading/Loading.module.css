@import 'styles/index';

:root {
  --loading-size-small: 5px;
  --loading-size-medium: 7px;
  --loading-size-large: 10px;
}

.base {
  display: flex;
  width: 100%;
  align-items: center;
  justify-content: center;
  pointer-events: none;
}

.small {
  height: var(--loading-size-small);
}

.medium {
  height: var(--loading-size-medium);
}

.large {
  height: var(--loading-size-large);
}

.fullSize {
  height: 100%;
}

.dot {
  display: block;
  margin: 0 3px;
  border-radius: 100%;

  .small & {
    width: var(--loading-size-small);
    height: var(--loading-size-small);
  }

  .medium & {
    width: var(--loading-size-medium);
    height: var(--loading-size-medium);
  }

  .large & {
    width: var(--loading-size-large);
    height: var(--loading-size-large);
  }

  .dark & {
    background: var(--color-grey-20);
  }

  .light & {
    background: var(--color-white);
  }

  @keyframes fadeInOut {
    0% {
      opacity: 0;
    }

    60% {
      opacity: 1;
    }

    100% {
      opacity: 0;
    }
  }

  .isLoading & {
    animation: fadeInOut 1000ms var(--easing-ease-in-out-cubic) infinite;

    &:nth-of-type(1) {
      animation-delay: -50ms;
    }

    &:nth-of-type(2) {
      animation-delay: 50ms;
    }

    &:nth-of-type(3) {
      animation-delay: 150ms;
    }
  }
}
