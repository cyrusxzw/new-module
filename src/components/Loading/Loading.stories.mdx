import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { StorybookGridOverlay } from '~.storybook/components';
import { Loading } from './Loading';

<Meta title="Components/Loading" component={Loading} />

# Loading

<Canvas>
  <Story name="Base component">
    {(() => {
      const isLoading = knobs.boolean('isLoading', true, 'Presentation');
      const shouldFillSpace = knobs.boolean(
        'shouldFillSpace',
        false,
        'Presentation',
      );
      const size = knobs.select(
        'size',
        ['small', 'medium', 'large'],
        'medium',
        'Presentation',
      );
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        'dark',
        'Presentation',
      );
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
      };
      return (
        <>
          <Loading
            isLoading={isLoading}
            shouldFillSpace={shouldFillSpace}
            size={size}
            theme={theme}
          />
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={Loading} />
