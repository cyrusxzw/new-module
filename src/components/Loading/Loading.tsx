import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import type { LoadingType } from './Loading.types';
import styles from './Loading.module.css';

const Loading: LoadingType = ({
  className,
  isLoading,
  shouldFillSpace = false,
  size = 'medium',
  theme,
}) => {
  const currentTheme = useThemeContext(theme, 'dark');

  if (!isLoading) return null;

  const classSet = cx(
    styles.base,
    {
      [styles.isLoading]: isLoading,
      [styles.fullSize]: shouldFillSpace,
    },
    styles[size],
    styles[currentTheme],
    className,
  );

  return (
    <span
      aria-busy={true}
      aria-live="polite"
      className={classSet}
      data-testid="data-testid-loading"
    >
      <span className={styles.dot} />
      <span className={styles.dot} />
      <span className={styles.dot} />
    </span>
  );
};

export { Loading };
