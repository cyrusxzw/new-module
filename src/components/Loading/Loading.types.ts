import type { ComponentWithoutChildren, Themes } from '~/types';

type LoadingSizes = 'small' | 'medium' | 'large';

type LoadingProps = {
  className?: string;
  isLoading: boolean;
  screenReaderText?: string;
  shouldFillSpace?: boolean;
  size?: LoadingSizes;
  theme?: Themes;
};

type LoadingType = ComponentWithoutChildren<LoadingProps>;

export type { LoadingProps, LoadingType };
