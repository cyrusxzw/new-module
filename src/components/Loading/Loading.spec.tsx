import React from 'react';
import { render } from '@testing-library/react';
import { Loading } from './Loading';

describe('<Loading />', () => {
  it('should render base component correctly', () => {
    const { container } = render(<Loading isLoading={true} />);

    expect(container).toMatchSnapshot();
  });
});
