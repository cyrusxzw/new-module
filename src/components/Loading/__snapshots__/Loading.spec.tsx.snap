// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render base component correctly 1`] = `
<div>
  <span
    aria-busy="true"
    aria-live="polite"
    class="base isLoading medium dark"
    data-testid="data-testid-loading"
  >
    <span
      class="dot"
    />
    <span
      class="dot"
    />
    <span
      class="dot"
    />
  </span>
</div>
`;
